"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.retry = void 0;
const errors_1 = require("./errors");
const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
const retry = async (routine) => {
    let retries = 3;
    let attempts = 0;
    while (attempts++ < retries) {
        try {
            return await routine();
        }
        catch (error) {
            if (error instanceof errors_1.CancelledCompletionError) {
                throw error;
            }
            if (retries === attempts) {
                throw error;
            }
            console.error(error);
            console.warn(`retrying after error: ${error.message}`);
        }
        await delay(1000);
    }
    throw new Error("Expected to never reach this point");
};
exports.retry = retry;
